### PR TITLE
schema rotation

### PR DESCRIPTION
**Pull request by benseligman**
Support configuration option to determine where to put writes (atomic per-write). This will allow applications to rotate schemas for redshift.
https://www.pivotaltracker.com/story/show/100840756

### CODE REVIEW

File: spec/support/backend.rb [hunk 1]
```diff
@@ -0,0 +1,78 @@
+module BackendHelper
```
Comment (benseligman) [inline spec/support/backend.rb]: mostly just copied out of promiscuous so we aren't coupled to promiscuous spec helpers

File: spec/support/define_constant.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
+require 'set'
+
+module DefineConstantMacros
```
Comment (benseligman) [inline spec/support/define_constant.rb]: just copied out of promiscuous so we aren't coupled to promiscuous spec helpers

File: spec/support/kafka.rb [hunk 3]
```diff
@@ -0,0 +1,63 @@
+module KafkaHelper
```
Comment (benseligman) [inline spec/support/kafka.rb]: just copied out of promiscuous so we aren't coupled to promiscuous spec helpers except for commented line

File: spec/support/test_cluster.rb [hunk 4]
```diff
@@ -0,0 +1,142 @@
+#taken from: https://github.com/bsm/promiscuous_cluster/blob/09241015f293bc6056e45af523dd2dce43643d21/scenario/scenario.rb
+
+require 'fileutils'
+require 'pathname'
+
+class TestCluster
```
Comment (benseligman) [inline spec/support/test_cluster.rb]: except for commented line, just copied out of promiscuous so we aren't coupled to promiscuous spec helpers

File: spec/support/test_cluster.rb [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        sub("=2181", "=#{ZOOKP_PORT}")
+    end
+    content = LOG4J_CFG.read
+    LOG4J_CFG.open("w") do |f|
+      f.write content.gsub("INFO", "FATAL")
+    end if content.include?("INFO")
+  end
+
+  def create_test_topic
+    sh "#{KAFKA_TOPIC} --zookeeper localhost:#{ZOOKP_PORT} --create --topic #{TOPIC} --partitions 2 --replication-factor 1"
```
Comment (benseligman) [inline spec/support/test_cluster.rb]: increase partitions, so we can do multi-threaded tests.

File: lib/promiscuous_black_hole/db.rb [hunk 6]
```diff
@@ -1,20 +1,76 @@
...
+    rescue => e
+      raise e
     end
 
-    def self.method_missing(method, *args, &block)
-      @@connection.public_send(method, *args, &block)
+    class << self
+      private
+      def set_schema_for_transaction(name)
+      end
```
Comment (michaeldauria) [inline lib/promiscuous_black_hole/db.rb]: Is this method being used?

File: spec/integration/schema_rotation_spec.rb [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+      eventually do
+        DB.transaction_with_applied_schema(@expected_schema_name) do
+          expect(DB[:publisher_models].count).to eq 1
+        end
+      end
+    end
+  end
+
+  it 'gracefully switches between schemata' do
```
Comment (michaeldauria) [inline spec/integration/schema_rotation_spec.rb]: Did you get this to consistently work?

File: spec/support/models.rb [hunk 8]
```diff
@@ -3,6 +3,7 @@ def load_models
     define_constant :PublisherModel do
       include Mongoid::Document
       include Promiscuous::Publisher
+      include Mongoid::Attributes::Dynamic
```
Comment (michaeldauria) [inline spec/support/models.rb]: this doesn't come with `Mongoid::Document`?

**Comment by michaeldauria** [comment]
:+1: otherwise

File: spec/support/models.rb (hunk 8 above)
Comment (benseligman) [inline spec/support/models.rb]: In mongoid 4, dynamic attributes are disabled by default and can be turned on per-collection with this method.

File: spec/integration/schema_rotation_spec.rb (hunk 7 above)
Comment (benseligman) [inline spec/integration/schema_rotation_spec.rb]: moving schema creation out of the transaction seems to fix it.

**Comment by iures** [comment]
lgtm!

**Pull request closed without merging**